### PR TITLE
[ZEPPELIN-2627] Interpreter Component Refactoring

### PR DESCRIPTION
**Pull request by zjffdu**
What is this PR for?

I didn't intended to make such large change at the beginning, but found many things are coupled together that I have to make such large change. Several suggestions for you how to review and read it.

I move the interpreter package from zeppelin-zengine to zeppelin-interpreter, this is needed for this refactoring.
The overall change is the same as I described in the design doc. I would suggest you to read the unit test first. These unit test is very readable and easy to understand what the code is doing now. `InterpreterFactoryTest`, `InterpreterGroupTest`, `InterpreterSettingTest`, `InterpreterSettingManagerTest`, `RemoteInterpreterTest`.
Remove the referent counting logic. Now I will kill the interpreter process as long as all the sessions in the same interpreter group is closed. (I plan to add another kind of policy for the interpreter process lifecycle control, ZEPPELIN-2197)
The RemoteFunction I introduced is for reducing code duplicates when we use RPC.
The changes in `Job.java` and `RemoteScheduler` is for fixing the race issue bug. This bug cause the flaky test we see often in `ZeppelinSparkClusterTest.pySparkTest`
What type of PR is it?

[Bug Fix | Refactoring]

Todos

 - Task
What is the Jira issue?

https://issues.apache.org/jira/browse/ZEPPELIN-2627
How should this be tested?

Unit test is added

Screenshots (if appropriate)

Questions:

Does the licenses files need update? No
Is there breaking changes for older versions? No
Does this needs documentation? No

### CODE REVIEW

**Comment by Leemoonsoo** [comment]
I can see some test cases are removed or commented out. for example zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterTest.java and some other places.

Is it intended or work in progress?

File: zeppelin-interpreter/src/test/java/org/apache/zeppelin/interpreter/InterpreterOutputChangeWatcherTest.java [hunk 1]
```diff
@@ -40,7 +41,7 @@ public void setUp() throws Exception {
     tmpDir = new File(System.getProperty("java.io.tmpdir")+"/ZeppelinLTest_"+System.currentTimeMillis());
     tmpDir.mkdirs();
     fileChanged = null;
-    numChanged = 0;
+    numChanged = new AtomicInteger(0);
```
Comment (zjffdu) [inline zeppelin-interpreter/src/test/java/org/apache/zeppelin/interpreter/InterpreterOutputChangeWatcherTest.java]: Fix flaky test

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterProcessTest.java [hunk 2]
```diff
@@ -1,131 +0,0 @@
...
-  private static final int DUMMY_PORT=3678;
-
-  @Test
-  public void testStartStop() {
-    InterpreterGroup intpGroup = new InterpreterGroup();
-    RemoteInterpreterManagedProcess rip = new RemoteInterpreterManagedProcess(
-        INTERPRETER_SCRIPT, "nonexists", "fakeRepo", new HashMap<String, String>(),
-        10 * 1000, null, null,"fakeName");
-    assertFalse(rip.isRunning());
-    assertEquals(0, rip.referenceCount());
```
Comment (zjffdu) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterProcessTest.java]: These kinds of testing is not needed as any more because of refactoring

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NoteInterpreterLoaderTest.java [hunk 3]
```diff
@@ -1,243 +0,0 @@
...
-public class NoteInterpreterLoaderTest {
-
-  private File tmpDir;
-  private ZeppelinConfiguration conf;
-  private InterpreterFactory factory;
-  private InterpreterSettingManager interpreterSettingManager;
-  private DependencyResolver depResolver;
-
-  @Before
-  public void setUp() throws Exception {
```
Comment (zjffdu) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NoteInterpreterLoaderTest.java]: All the test here is covered by `InterpreterSettingTest` & `InterpreterSettingManagerTest`

**Comment by zjffdu** [comment]
@Leemoonsoo I add more comments for the test case. Some testcase is not needed any more

**Comment by zjffdu** [comment]
@Leemoonsoo Do you mind if I commit it first to continue other works ?

**Comment by Leemoonsoo** [comment]
Let me test this branch little more with some edge cases.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/scheduler/RemoteScheduler.java [hunk 4]
```diff
@@ -338,20 +317,22 @@ public void run() {
...
-        if (((InterpreterResult) jobResult).code() == Code.ERROR) {
-          lastStatus = Status.ERROR;
-        }
-      }
-      if (job.getException() != null) {
-        lastStatus = Status.ERROR;
+      if (job.isAborted()) {
+        job.setStatus(Status.ABORT);
+      } else if (job.getException() != null) {
+//        logger.info("Job ABORT, " + job.getId());
```
Comment (Leemoonsoo) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/scheduler/RemoteScheduler.java]: Shell we remove or change to logger.debug?

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java [hunk 5]
```diff
@@ -17,481 +17,50 @@
...
+    interpreterFactory.getInterpreter("user1", "note1", "test.unknown_repl");
   }
 
-  /**
-   * 2 users' interpreters in scoped mode. Each user has one session. Restarting user1's interpreter
-   * won't affect user2's interpreter
-   * @throws Exception
-   */
   @Test
-  public void testRestartInterpreterInScopedMode() throws Exception {
```
Comment (Leemoonsoo) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: @zjffdu Could you point new testcase covers this case?
InterpreterSettingManagerTest looks like have related testcase but somehow restart behavior in scoped mode seems changed. try

1. set an interpreter scoped mode per note
2. create two notes and start a paragraph in each note 
3. Restart intepreter from a note in 'interpreter-binding' menu (not in interpreter menu)

Interpreter supposed to be restart for a note only, but it restarts interpreter for both notes.

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java [hunk 6]
```diff
@@ -17,481 +17,50 @@
...
-    assertTrue(interpreter2.isOpen());
-  }
-
-  /**
-   * 2 users' interpreters in isolated mode. Each user has one interpreterGroup. Restarting user1's interpreter
-   * won't affect user2's interpreter
-   * @throws Exception
-   */
-  @Test
-  public void testRestartInterpreterInIsolatedMode() throws Exception {
```
Comment (Leemoonsoo) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: Just like `testRestartInterpreterInScopedMode()` restart behavior protected by this unittest has changed.
We'll need to bring this testcase back.

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java [hunk 7]
```diff
@@ -17,481 +17,50 @@
...
-      add(setting1.getId());
-      add(setting2.getId());
-    }});
-
-    assertEquals("className1", factory.getInterpreter("user1", "note", "test-group1").getClassName());
-    assertEquals("className1", factory.getInterpreter("user1", "note", "group1").getClassName());
-  }
-
-  @Test
-  public void testMultiUser() throws IOException, RepositoryException {
```
Comment (Leemoonsoo) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: Do we have an equivalent test for this?

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java [hunk 8]
```diff
@@ -17,481 +17,50 @@
...
-    editor = interpreterSettingManager.getEditorSetting(factory.getInterpreter("user1", note.getId(), "mock1"),"user1", note.getId(), "mock1");
-    assertEquals(null, editor.get("language"));
-
-    // when interpreter is not bound to note
-    editor = interpreterSettingManager.getEditorSetting(factory.getInterpreter("user1", note.getId(), "mock11"),"user1", note.getId(), "mock2");
-    assertEquals("text", editor.get("language"));
-  }
-
-  @Test
-  public void registerCustomInterpreterRunner() throws IOException {
```
Comment (Leemoonsoo) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: Do we have an equivalent test for this?

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java [hunk 9]
```diff
@@ -17,481 +17,50 @@
...
-      interpreterSettingManager.createNewSetting("new.mock1", "mock1", new LinkedList<Dependency>(), new InterpreterOption(false), new HashMap<String, InterpreterProperty>());
-      fail("expect fail because of invalid InterpreterSetting Name");
-    } catch (IOException e) {
-      assertEquals("'.' is invalid for InterpreterSetting name.", e.getMessage());
-    }
-  }
-
-
-  @Test
-  public void getEditorSetting() throws IOException, RepositoryException, SchedulerException {
```
Comment (Leemoonsoo) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: Do we have an equivalent test for this?

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java (hunk 5 above)
Comment (zjffdu) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: See test here https://github.com/apache/zeppelin/pull/2554/files#diff-6e873dfc1f240d8453d17a98f2ae4d10R180
and 
https://github.com/apache/zeppelin/pull/2554/files#diff-63ebbe353cb86fc2eb25d29e14e7cb18R101

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java (hunk 6 above)
Comment (zjffdu) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: https://github.com/apache/zeppelin/pull/2554/files#diff-63ebbe353cb86fc2eb25d29e14e7cb18R175

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java (hunk 7 above)
Comment (zjffdu) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: I think it is covered by https://github.com/apache/zeppelin/pull/2554/files#diff-63ebbe353cb86fc2eb25d29e14e7cb18R175

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java (hunk 8 above)
Comment (zjffdu) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: It is covered here. https://github.com/apache/zeppelin/pull/2554/files#diff-6e873dfc1f240d8453d17a98f2ae4d10R69

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java (hunk 9 above)
Comment (zjffdu) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: It is covered here https://github.com/apache/zeppelin/pull/2554/files#diff-6e873dfc1f240d8453d17a98f2ae4d10R273

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java (hunk 5 above)
Comment (Leemoonsoo) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: Yes, those tests suppose to ensure restart behavior from note page.
However, if I try this by hands, I'm getting different behavior explained above.

Could you take a look?

Comment (zjffdu) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: @Leemoonsoo I believe this is the current behavior, not caused by this PR. Not sure why we would close all interpreter when it is `anonymous`. 
https://github.com/apache/zeppelin/blob/master/zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java#L971

Comment (Leemoonsoo) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java]: You're right. I have tested master branch with shiro configured and this branch without shiro.
I also not sure why close all interpreter when it is `anonymous`.

It's up to you change the behavior in this PR or not, but maybe handle it in separate PR for future reference?

**Comment by Leemoonsoo** [comment]
Thanks @zjffdu for great work.
LGTM

**Comment by zjffdu** [comment]
Thanks @Leemoonsoo for the review, I will merge it if no more comments. And will do the change in the following PR as there's still remaining work for the refactoring.

**Comment by prabhjyotsingh** [comment]
@zjffdu  Just found out an issue which is a side effect of this PR, now I cannot delete an interpreter. 
By that what I mean is now say if I delete angular interpreter from the interpreter setting page, and restart zeppelin it gets recreated.

**Comment by zjffdu** [comment]
hmm, I see. Because every time I will merge interpreter-setting.json into interpreter-setting.json. Mind to create a ticket ? I will fix it.

**Comment by prabhjyotsingh** [comment]
Sure here you go [ZEPPELIN-3029](https://issues.apache.org/jira/browse/ZEPPELIN-3029)
